hypothesis-python/docs/tutorial/more-strategies.rst

```diff
@@ -0,0 +1,264 @@
+
+If we run this test, we get an error:
+
+.. code-block:: none
+
+        ...
+        return [f / sum(lst) for f in lst]
+                ~~^~~~~~~~~~
+    ZeroDivisionError: float division by zero
+    while generating 'lst' from sums_to_one()
+
+Whoops. We forgot that we could generate lists filled with just zero. This case seems difficult to avoid by construction without sacrificing test power, so we can filter it out with |assume|:
+
+.. code-block:: python
+
+    from hypothesis import strategies as st, given, assume
+    import pytest
+
+    @st.composite
+    def sums_to_one(draw):
+        lst = draw(st.lists(st.floats(0, 1), min_size=1))
+        assume(sum(lst) > 0)  #  <-- new
+        return [f / sum(lst) for f in lst]
+
+    @given(sums_to_one())
+    def test(lst):
+        # ignore floating point errors
+        assert sum(lst) == pytest.approx(1)
+
+|assume| works exactly the same when called inside |st.composite| as it does when called in a test.
+
+Parameters with |st.composite|
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can add parameters to functions decorated with |st.composite|, including keyword-only arguments. These work as you would normally expect.
+
+For instance, suppose we wanted to generalize our ``sums_to_one`` function to ``sums_to_n``. We can add a parameter ``n``:
+
+.. code-block:: python
+
+    from hypothesis import strategies as st, given, assume
+    import pytest
+
+    @st.composite
+    def sums_to_n(draw, n=1):  #  <-- changed
+        lst = draw(st.lists(st.floats(0, 1), min_size=1))
+        assume(sum(lst) > 0)
+        return [f / sum(lst) * n for f in lst]  #  <-- changed
+
+    @given(sums_to_n(10))
+    def test(lst):
+        assert sum(lst) == pytest.approx(10)
+
+And we could just as easily have made ``n`` a keyword-only argument instead:
+
+.. code-block:: python
+
+    from hypothesis import strategies as st, given, assume
+    import pytest
+
+    @st.composite
+    def sums_to_n(draw, *, n=1):  #  <-- changed
+        lst = draw(st.lists(st.floats(0, 1), min_size=1))
+        assume(sum(lst) > 0)
+        return [f / sum(lst) * n for f in lst]
+
+    @given(sums_to_n(n=10))  #  <-- changed
+    def test(lst):
+        assert sum(lst) == pytest.approx(10)
+
+Dependent generation
+~~~~~~~~~~~~~~~~~~~~
+
+Another scenario where |st.composite| is useful is when generating a value that depends on a value from another strategy. For instance, suppose we wanted to generate two integers ``n1`` and ``n2`` with ``n1 <= n2``. We can do this using |st.composite|:
+
+.. code-block:: python
+
+    @st.composite
+    def integer_pairs(draw):
+        n1 = draw(st.integers())
+        n2 = draw(st.integers(min_value=n1))
+        return (n1, n2)
+
+    @given(integer_pairs())
+    def test_integer_pairs(pair):
+        n1, n2 = pair
+        assert n1 <= n2
```

Comment:
Afterwards, I'd add

> Consider that that you could also use `st.tuples(...).map(sorted)` for the particular case of an ordered pair.  Some people prefer creatively composing strategy methods, others prefer well-named helper functions.  Our suggestion is simply that you prioritize ease of understanding when choosing in each case.

(which in fact suggests that the helper should be named `ordered_integer_pairs`)